### PR TITLE
feat: Config and Argument configuration plus tests for these features

### LINKED ISSUE #438: Chart CI to identify issues before merging
#436 was caused by a typo that I believe should have been picked up with the tooling I tested with.

For example I passed the template through [kubeconform](https://github.com/yannh/kubeconform):

```shell
helm template saas oci://${TLSPK_REGISTRY_EU}/charts/jetstack-agent \
--registry-config "${TLSPK_DOCKER_CONFIG_FILE}" \
--namespace jetstack-agent-saas \
--version 0.2.0 \
--values ~/temp/js-agent-saas-values.yaml | kubeconform -summary -verbose

Pulled: eu.gcr.io/jetstack-secure-enterprise/charts/jetstack-agent:0.2.0
Digest: sha256:a09dfed578c5fe580fcae026fb6596e3e8eafe6288024000bb93cb4b9afff8eb
stdin - ConfigMap agent-config is valid
stdin - ClusterRole jetstack-agent-saas-node-reader is valid
stdin - ClusterRole jetstack-agent-saas-cert-manager-reader is valid
stdin - ClusterRole jetstack-agent-saas-googlecas-reader is valid
stdin - ClusterRole jetstack-agent-saas-awspca-reader is valid
stdin - ClusterRole jetstack-agent-saas-webhook-reader is valid
stdin - ClusterRole jetstack-agent-saas-openshift-reader is valid
stdin - ClusterRole jetstack-agent-saas-istio-reader is valid
stdin - ClusterRole jetstack-agent-saas-venafi-enhanced-reader is valid
stdin - ServiceAccount jetstack-agent-saas is valid
stdin - Secret agent-credentials is valid
stdin - ClusterRole jetstack-agent-saas-secret-reader is valid
stdin - ClusterRoleBinding jetstack-agent-saas-cluster-viewer is valid
stdin - ClusterRoleBinding jetstack-agent-saas-googlecas-reader is valid
stdin - ClusterRoleBinding jetstack-agent-saas-awspca-reader is valid
stdin - ClusterRoleBinding jetstack-agent-saas-webhook-reader is valid
stdin - ClusterRoleBinding jetstack-agent-saas-openshift-reader is valid
stdin - ClusterRoleBinding jetstack-agent-saas-secret-reader is valid
stdin - ClusterRoleBinding jetstack-agent-saas-istio-reader is valid
stdin - ClusterRoleBinding jetstack-agent-saas-node-reader is valid
stdin - ClusterRoleBinding jetstack-agent-saas-venafi-enhanced-reader is valid
stdin - ClusterRoleBinding jetstack-agent-saas-cert-manager-reader is valid
stdin - Deployment jetstack-agent-saas is valid
Summary: 23 resources found parsing stdin - Valid: 23, Invalid: 0, Errors: 0, Skipped: 0
```

It wasn't until the actual files are applied to k8s that the issue was seen:

```
helm template saas oci://${TLSPK_REGISTRY_EU}/charts/jetstack-agent \
--registry-config "${TLSPK_DOCKER_CONFIG_FILE}" \
--namespace jetstack-agent-saas \
--version 0.2.0 \
--values ~/temp/js-agent-saas-values.yaml | kubectl apply -f - --dry-run=server
Pulled: eu.gcr.io/jetstack-secure-enterprise/charts/jetstack-agent:0.2.0
Digest: sha256:a09dfed578c5fe580fcae026fb6596e3e8eafe6288024000bb93cb4b9afff8eb
serviceaccount/jetstack-agent-saas created (server dry run)
secret/agent-credentials configured (server dry run)
Warning: resource configmaps/agent-config is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by kubectl apply. kubectl apply should only be used on resources created declaratively by either kubectl create --save-config or kubectl apply. The missing annotation will be patched automatically.
configmap/agent-config configured (server dry run)
clusterrole.rbac.authorization.k8s.io/jetstack-agent-saas-node-reader created (server dry run)
clusterrole.rbac.authorization.k8s.io/jetstack-agent-saas-secret-reader created (server dry run)
clusterrole.rbac.authorization.k8s.io/jetstack-agent-saas-cert-manager-reader created (server dry run)
clusterrole.rbac.authorization.k8s.io/jetstack-agent-saas-googlecas-reader created (server dry run)
clusterrole.rbac.authorization.k8s.io/jetstack-agent-saas-awspca-reader created (server dry run)
clusterrole.rbac.authorization.k8s.io/jetstack-agent-saas-webhook-reader created (server dry run)
clusterrole.rbac.authorization.k8s.io/jetstack-agent-saas-openshift-reader created (server dry run)
clusterrole.rbac.authorization.k8s.io/jetstack-agent-saas-istio-reader created (server dry run)
clusterrole.rbac.authorization.k8s.io/jetstack-agent-saas-venafi-enhanced-reader created (server dry run)
clusterrolebinding.rbac.authorization.k8s.io/jetstack-agent-saas-cluster-viewer created (server dry run)
clusterrolebinding.rbac.authorization.k8s.io/jetstack-agent-saas-node-reader created (server dry run)
clusterrolebinding.rbac.authorization.k8s.io/jetstack-agent-saas-secret-reader created (server dry run)
clusterrolebinding.rbac.authorization.k8s.io/jetstack-agent-saas-cert-manager-reader created (server dry run)
clusterrolebinding.rbac.authorization.k8s.io/jetstack-agent-saas-googlecas-reader created (server dry run)
clusterrolebinding.rbac.authorization.k8s.io/jetstack-agent-saas-awspca-reader created (server dry run)
clusterrolebinding.rbac.authorization.k8s.io/jetstack-agent-saas-webhook-reader created (server dry run)
clusterrolebinding.rbac.authorization.k8s.io/jetstack-agent-saas-openshift-reader created (server dry run)
clusterrolebinding.rbac.authorization.k8s.io/jetstack-agent-saas-istio-reader created (server dry run)
clusterrolebinding.rbac.authorization.k8s.io/jetstack-agent-saas-venafi-enhanced-reader created (server dry run)
Error from server (BadRequest): error when creating "STDIN": Deployment in version "v1" cannot be handled as a Deployment: strict decoding error: unknown field "spec.template.spec.volumes[0].configMap.options"
```

In order to prevent an issue like this in the future we should probably have some CI tooling to perform some testing on the chart. Not sure what the best tooling option is for this, so opening this issues to discuss.

Thinking that using a linting tool like `kubeconform` is not enough, but actually installing the chart in a kind environment and checking for errors would be a better validation.

### PR DESCRIPTION
**Pull request by hawksight**
Helps to resolve #438 as it adds [helm unittest](https://github.com/helm-unittest/helm-unittest) to test the new features added.

Allows:
- Passing `extraArgs` to the agent.
- Allows you to embed the agent config as completely custom in the chart
- Allows you to specify a completely different configmap (& key) so config is not within the chart.
- Update to v0.1.40 of agent.
- CronJobs to `v1` not `v1beta1`

### CODE REVIEW

**Comment by hawksight** [comment]
Kind of a little stuck on ensuring the actual config.yaml file is templated correctly.
See [this issue](https://github.com/helm-unittest/helm-unittest/issues/213) for context.

**Comment by hawksight** [comment]
I think this is ready for a review now. Appreciate a good set of eyes on this as I've bumped to v0.3.0 as there's a fair few changes around config options here.

File: deploy/charts/jetstack-agent/README.md [hunk 1]
```diff
@@ -137,19 +137,24 @@ kubectl logs -n jetstack-secure $(kubectl get pod -n jetstack-secure -l app.kube
...
 | config.cluster | string | `""` | REQUIRED - Your Jetstack Secure Cluster Name |
 | config.dataGatherers | object | `{"custom":[],"default":true}` | Configure data that is gathered from your cluster, for full details see https://platform.jetstack.io/documentation/configuration/jetstack-agent/configuration |
 | config.dataGatherers.custom | list | `[]` | A list of data gatherers to limit agent scope |
 | config.dataGatherers.default | bool | `true` | Use the standard full set of data gatherers |
 | config.organisation | string | `""` | REQUIRED - Your Jetstack Secure Organisation Name |
+| config.override | object | `{"config":"","configmap":{"key":"","name":""},"enabled":false}` | Provide an Override to allow completely custom agent configuration |
+| config.override.config | string | `""` | Embed the agent configuration here in the chart values |
+| config.override.configmap | object | `{"key":"","name":""}` | Sepcify ConfigMap details to load config from existing ConfigMap |
+| config.override.enabled | bool | `false` | Override disabled by default |
```
Comment (SpectralHiss) [inline deploy/charts/jetstack-agent/README.md:151]: Could we make this more implicit? like set the config.override to null and if it has set config or configmap it is automatically "enabled"? just a small UX preference for me, i like to use enabled but for bigger pieces of config if that makes sense..

File: deploy/charts/jetstack-agent/templates/configmap.yaml [hunk 2]
```diff
@@ -6,8 +7,12 @@ metadata:
   labels:
     {{- include "jetstack-agent.labels" . | nindent 4 }}
 data:
+{{- if .Values.config.override.config }}
```
Comment (SpectralHiss) [inline deploy/charts/jetstack-agent/templates/configmap.yaml:10]: This could just be a nullity check for example (cf suggestion on the values README)

**Review by SpectralHiss — CHANGES_REQUESTED** [review]
Only 1 minor suggestion to change the override "interface" a little bit.

File: deploy/charts/jetstack-agent/README.md [hunk 3]
```diff
@@ -137,19 +137,24 @@ kubectl logs -n jetstack-secure $(kubectl get pod -n jetstack-secure -l app.kube
 | authentication.secretKey | string | `"credentials.json"` | Key name in secret |
 | authentication.secretName | string | `"agent-credentials"` | Name of the secret containing agent credentials.json |
 | authentication.secretValue | string | `""` | Base64 encoded value from Jetstack Secure Dashboard - only required when createSecret is true |
-| authentication.type | string | `"file"` | Type can be "file"/"token" which dictates whether a credential file or token is used |
-| config | object | `{"cluster":"","dataGatherers":{"custom":[],"default":true},"organisation":"","period":"0h1m0s","server":"https://platform.jetstack.io"}` | Configuration section for the Jetstack Agent itself |
+| authentication.type | string | `"file"` | Type can be "file"/"token" determining how the agent should authenticate the to the backend |
+| command | list | `[]` |  |
+| config | object | `{"cluster":"","dataGatherers":{"custom":[],"default":true},"organisation":"","override":{"config":"","configmap":{"key":"","name":""},"enabled":false},"period":"0h1m0s","server":"https://platform.jetstack.io"}` | Configuration section for the Jetstack Agent itself |
```
[the remark this answers is not in the document]
Comment (hawksight) [reply deploy/charts/jetstack-agent/README.md:142]: Yes probably - will rename for the next chart.

File: deploy/charts/jetstack-agent/README.md (hunk 1 above)
Comment (hawksight) [reply deploy/charts/jetstack-agent/README.md:151]: We explored the option and found it's just too dirty in helm templating to make this work nicely.

**Comment by hawksight** [comment]
Had a huddle with @SpectralHiss to see if we could make the chart simpler with:

```yaml
  override:
    # -- Embed the agent configuration here in the chart values
    config: 
    # -- Sepcify ConfigMap details to load config from existing ConfigMap
    configmap:
      name: 
      key: 
```

Or even simpler with a default of:

```yaml
  override: {}
```

However it turned out to be too difficult to make the logic cleaner in the chart.
So for now we agreed to leave how it is and improve for the next version of the chart that will be Venafi Specific

**Review by SpectralHiss — APPROVED** [review]

**Review by j-fuentes — APPROVED** [review]

**Pull request merged**